### PR TITLE
NIFI-13123 MiNiFi async C2 operation processing

### PR DESCRIPTION
**Pull request by briansolo1985**
# Summary

[NIFI-13123](https://issues.apache.org/jira/browse/NIFI-13123)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### UI Contributions

- [ ] NiFi is modernizing its UI. Any contributions that update the [current UI](https://github.com/apache/nifi/tree/main/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui) also need to be implemented in the [new UI](https://github.com/apache/nifi/tree/main/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi).  

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-framework-api/src/main/java/org/apache/nifi/bootstrap/BootstrapCommunicator.java [hunk 1]
```diff
@@ -25,11 +25,13 @@ public interface BootstrapCommunicator {
...
      * Sends a command with specific arguments to the bootstrap process
+     *
      * @param command the command to send
-     * @param args the args to send
+     * @param args    the args to send
+     * @return whether the command send was successful or not
      * @throws IOException exception in case of communication issue
      */
-    void sendCommand(String command, String... args) throws IOException;
+    boolean sendCommand(String command, String... args) throws IOException;
```
Comment (exceptionfactory) [inline nifi-framework-api/src/main/java/org/apache/nifi/bootstrap/BootstrapCommunicator.java:34]: For future flexibility, I recommend changing the `boolean` to an `enum` named something like `CommandResult` or `CommandStatus`. The initial values could be `FAILURE` and `SUCCESS`, but using an `enum` provides more options for extension with changing the interface contract.

File: minifi/minifi-commons/minifi-commons-api/src/main/java/org/apache/nifi/minifi/commons/api/MiNiFiProperties.java [hunk 2]
```diff
@@ -91,6 +91,7 @@ public enum MiNiFiProperties {
     C2_SECURITY_KEYSTORE_PASSWORD("c2.security.keystore.password", "", true, false, VALID),
     C2_SECURITY_KEYSTORE_TYPE("c2.security.keystore.type", "JKS", false, false, VALID),
     C2_REQUEST_COMPRESSION("c2.request.compression", "none", false, true, VALID),
+    C2_BOOTSTRAP_ACKNOWLEDGE_TIMEOUT("c2.bootstrap.acknowledge.timeout", "60 sec", false, true, VALID),
```
Comment (bejancsaba) [inline minifi/minifi-commons/minifi-commons-api/src/main/java/org/apache/nifi/minifi/commons/api/MiNiFiProperties.java:94]: This seems like a lot for a default if I understand the use case right but maybe there are arguments against lowering this number. Just wanted to raise I thought 10-15sec should be sufficient or for complex flows restart could take really long?

**Review by bejancsaba — APPROVED** [review]
I went through the code and tried that operation processing works as expected. At first round I was confused because I didn't find the logic that processes the operations in an async manner but I realised this was about decoupling the operation processing from heartbeating and not the parallelisation of the operation processing itself (the ticcket explained it clearly). Anyway the code while complex is very well structured, it waas easy to follow. The test coverage gives good confidence and as manually tested the behaviour is as expected. I had just minor notes so even if you don't change those I'm +1. Thanks for the change also around cleaning up how the network information is determined.

File: nifi-framework-api/src/main/java/org/apache/nifi/bootstrap/BootstrapCommunicator.java (hunk 1 above)
Comment (briansolo1985) [reply nifi-framework-api/src/main/java/org/apache/nifi/bootstrap/BootstrapCommunicator.java]: Fixed

File: c2/c2-client-bundle/c2-client-service/src/main/java/org/apache/nifi/c2/client/service/C2OperationManager.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                operation = c2Operations.take();
+            } catch (InterruptedException e) {
+                LOGGER.warn("Thread was interrupted", e);
+                return;
+            }
+
+            LOGGER.debug("Processing operation {}", operation);
+            C2OperationHandler operationHandler = c2OperationHandlerProvider.getHandlerForOperation(operation).orElse(null);
+            if (operationHandler == null) {
+                LOGGER.debug("No handler is present for for C2 Operation {}", operation);
```
[the remark this answers is not in the document]
Comment (briansolo1985) [reply c2/c2-client-bundle/c2-client-service/src/main/java/org/apache/nifi/c2/client/service/C2OperationManager.java]: Removed

File: c2/c2-client-bundle/c2-client-service/src/test/java/org/apache/nifi/c2/client/service/C2HeartbeatManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    @Test
+    void shouldSendHeartbeatAndProcessResponseWithNoOperation() {
+        when(mockHeartbeatLock.tryLock()).thenReturn(true);
+        C2Heartbeat mockC2Heartbeat = mock(C2Heartbeat.class);
+        when(mockC2HeartbeatFactory.create(mockRuntimeInfoWrapper)).thenReturn(mockC2Heartbeat);
+        C2HeartbeatResponse mockC2HeartbeatResponse = mock(C2HeartbeatResponse.class);
+        when(mockC2HeartbeatResponse.getRequestedOperations()).thenReturn(List.of());
+        when(mockC2Client.publishHeartbeat(mockC2Heartbeat)).thenReturn(ofNullable(mockC2HeartbeatResponse));
```
[the remark this answers is not in the document]
Comment (briansolo1985) [reply c2/c2-client-bundle/c2-client-service/src/test/java/org/apache/nifi/c2/client/service/C2HeartbeatManagerTest.java]: Yes, fixed

File: minifi/minifi-commons/minifi-commons-api/src/main/java/org/apache/nifi/minifi/commons/api/MiNiFiProperties.java (hunk 2 above)
Comment (briansolo1985) [reply minifi/minifi-commons/minifi-commons-api/src/main/java/org/apache/nifi/minifi/commons/api/MiNiFiProperties.java]: Lowered to 15 secs

**Comment by briansolo1985** [comment]
Thanks for the review comments @exceptionfactory and @bejancsaba. Incorporated all of them, please review again

**Comment by ferencerdei** [comment]
Tested the change, works properly. +1 From my side

**Review by exceptionfactory — COMMENTED** [review]
Thanks for updating the `sendCommand` return type, @briansolo1985, that looks good.

**Review by bejancsaba — APPROVED** [review]
Thanks for the changes, all looks good +1

**Pull request closed without merging**